homeassistant/components/mqtt/discovery.py

```diff
@@ -155,116 +315,125 @@ def async_discovery_message_received(msg: ReceiveMessage) -> None:  # noqa: C901
 
         component, node_id, object_id = match.groups()
 
-        if component not in SUPPORTED_COMPONENTS:
-            _LOGGER.warning("Integration %s is not supported", component)
-            return
+        discovered_components: list[MqttComponentConfig] = []
+        if component == CONF_DEVICE:
+            # Process device based discovery message
+            device_discovery_payload = _parse_device_payload(
+                hass, payload, object_id, node_id
+            )
+            if not device_discovery_payload:
+                return
+            device_config: dict[str, Any] = device_discovery_payload[CONF_DEVICE]
+            origin_config: dict[str, Any] | None = device_discovery_payload.get(
+                CONF_ORIGIN
+            )
+            component_configs: dict[str, Any] = device_discovery_payload[
+                CONF_COMPONENTS
+            ]
+            for component_id, config in component_configs.items():
+                component = config.pop(CONF_PLATFORM)
+                # The object_id in the device discovery topic is the unique identifier.
+                # It is used as node_id for the components it contains.
+                component_node_id = object_id
+                # The component_id in the discovery playload is used as object_id
+                # If we have an additional node_id in the discovery topic,
+                # we extend the component_id with it.
+                component_object_id = (
+                    f"{node_id} {component_id}" if node_id else component_id
+                )
+                _replace_all_abbreviations(config)
+                discovery_payload = MQTTDiscoveryPayload(config)
+                if discovery_payload:
```

Comment:
It looks like this `if` can be dropped

Reply:
discovery_payload can be an empty dict forwarding removal of an entity. Still we apply MQTTDiscoveryPayload so we have the following attributes added:
```python
    device_discovery: bool = False
    discovery_data: DiscoveryInfoType
```

Reply:
Added extra comment to explain the dict can be empty and we still use the attributes in the wrapper.